### PR TITLE
rustdoc-search: clean up type unification and "unboxing"

### PR DESCRIPTION
**Pull request by notriddle**
This PR redesigns parameter matching, return matching, and generics matching to use a single function that compares two lists of types.

It also makes the algorithms more consistent, so the "unboxing" behavior where `Vec<i32>` is considered a match for `i32` works inside generics, and not just at the top level.

### CODE REVIEW

File: src/librustdoc/html/static/js/search.js [hunk 1]
```diff
@@ -1103,98 +1103,141 @@ function initSearch(rawSearchIndex) {
...
+                        continue unboxing;
                     }
+                    i += 1;
                 }
-                for (const generic of elem.generics) {
-                    if (generic.typeFilter === -1 && !handleGeneric(generic)) {
-                        return false;
+                for (const [id, fnTypeList] of fnTypeSet) {
+                    i = 0;
+                    for (const fnType of fnTypeList) {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1211]: Same here.

**Comment by GuillaumeGomez** [comment]
Apart from my comments, looks good to me: code well documented and changes make sense.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks! Just to confirm: no behaviour change in the search, right? If so r=me, otherwise please start an FCP.

**Comment by GuillaumeGomez** [comment]
It's fine but since it's a user facing change, FCP is needed I think. If not, we can always cancel it and at least the rest of the team will be able to be aware of this change.

@rfcbot fcp merge

**Comment by Manishearth** [comment]
Don't think fcp is needed since it feels almost like a bugfix (it's a minor improvement to functionality in a way that is obviously better). But sgtm

**Pull request merged**